File: spring/src/test/java/cardsproject/controller/content/StreamControllerTest.java
package cardsproject.controller.content;

import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.http.MediaType;
import org.springframework.test.web.servlet.MockMvc;
import org.springframework.transaction.annotation.Transactional;

import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.*;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.*;

@SpringBootTest
@AutoConfigureMockMvc
@Transactional
public class StreamControllerTest {

    @Autowired
    private MockMvc mockMvc;

    @Test
    void list_returns200() throws Exception {
        mockMvc.perform(get("/api/streams"))
            .andExpect(status().isOk());
    }
    @Test
    void search_returns200() throws Exception {
        mockMvc.perform(get("/api/streams?q=test"))
            .andExpect(status().isOk());
    }
    @Test
    void create_returns201() throws Exception {
        mockMvc.perform(post("/api/streams")
            .contentType(MediaType.APPLICATION_JSON)
            .content("{ \"title\": \"test\", \"streamUrl\": \"https://example.com\", \"scheduledStart\": \"2024-01-01T00:00:00\", \"actualStart\": null, \"endedAt\": null }"))
            .andExpect(status().isCreated());
    }
    @Test
    void show_returns200or404() throws Exception {
        mockMvc.perform(get("/api/streams/1"))
            .andExpect(result -> {
                int status = result.getResponse().getStatus();
                assert status == 200 || status == 404;
            });
    }
    @Test
    void update_returns200or404() throws Exception {
        mockMvc.perform(put("/api/streams/1")
            .contentType(MediaType.APPLICATION_JSON)
            .content("{ \"title\": \"test\", \"streamUrl\": \"https://example.com\", \"scheduledStart\": \"2024-01-01T00:00:00\", \"actualStart\": null, \"endedAt\": null }"))
            .andExpect(result -> {
                int status = result.getResponse().getStatus();
                assert status == 200 || status == 404;
            });
    }
    @Test
    void patch_returns200or404() throws Exception {
        mockMvc.perform(patch("/api/streams/1")
            .contentType(MediaType.APPLICATION_JSON)
            .content("{}"))
            .andExpect(result -> {
                int status = result.getResponse().getStatus();
                assert status == 200 || status == 404;
            });
    }
    @Test
    void create_fails_when_actual_start_requires_live_or_ended_violated() throws Exception {
        // actual_start_requires_live_or_ended: antecedent true, consequent missing → 400
        mockMvc.perform(post("/api/streams")
            .contentType(MediaType.APPLICATION_JSON)
            .content("{ \"title\": \"test\", \"streamUrl\": \"https://example.com\", \"status\": \"SCHEDULED\", \"platform\": \"TWITCH\", \"language\": \"EN\", \"isOfficial\": true, \"viewerCountPeak\": 1, \"scheduledStart\": \"2024-01-01T00:00:00\", \"streamerId\": 1, \"actualStart\": \"2024-01-01T00:00:00\" }"))
            .andExpect(status().isBadRequest());
    }

    @Test
    void create_fails_when_ended_at_requires_ended_status_violated() throws Exception {
        // ended_at can only be set when stream status is Ended: antecedent true, consequent missing → 400
        mockMvc.perform(post("/api/streams")
            .contentType(MediaType.APPLICATION_JSON)
            .content("{ \"title\": \"test\", \"streamUrl\": \"https://example.com\", \"status\": \"SCHEDULED\", \"platform\": \"TWITCH\", \"language\": \"EN\", \"isOfficial\": true, \"viewerCountPeak\": 1, \"scheduledStart\": \"2024-01-01T00:00:00\", \"streamerId\": 1, \"endedAt\": \"2024-01-01T00:00:00\" }"))
            .andExpect(status().isBadRequest());
    }

    @Test
    void create_fails_when_viewer_count_not_negative_violated() throws Exception {
        // Peak viewer count must not be negative → 400 (Bean Validation)
        mockMvc.perform(post("/api/streams")
            .contentType(MediaType.APPLICATION_JSON)
            .content("{ \"title\": \"test\", \"streamUrl\": \"https://example.com\", \"platform\": \"TWITCH\", \"language\": \"EN\", \"isOfficial\": true, \"scheduledStart\": \"2024-01-01T00:00:00\", \"streamerId\": 1, \"actualStart\": \"2024-01-01T00:00:00\", \"status\": \"LIVE\", \"endedAt\": \"2024-01-01T00:00:00\", \"status\": \"ENDED\", \"viewerCountPeak\": -1 }"))
            .andExpect(status().isBadRequest());
    }
    @Test
    @org.springframework.security.test.context.support.WithMockUser(roles = {"STREAMER"})
    void transitionScheduledToLive_returns200or404() throws Exception {
        mockMvc.perform(patch("/api/streams/1/transitions/scheduled-to-live"))
            .andExpect(result -> {
                int s = result.getResponse().getStatus();
                assert s == 200 || s == 404 || s == 409 || s == 422;
            });
    }

    @Test
    @org.springframework.security.test.context.support.WithMockUser(roles = {"STREAMER"})
    void transitionLiveToEnded_returns200or404() throws Exception {
        mockMvc.perform(patch("/api/streams/1/transitions/live-to-ended"))
            .andExpect(result -> {
                int s = result.getResponse().getStatus();
                assert s == 200 || s == 404 || s == 409 || s == 422;
            });
    }

    @Test
    void transitionEndedToLive_isDenied() throws Exception {
        mockMvc.perform(patch("/api/streams/1/transitions/ended-to-live"))
            .andExpect(result -> {
                int s = result.getResponse().getStatus();
                assert s == 409 || s == 404;
            });
    }
}

File: phoenix/test/content/stream_controller_test.exs
defmodule CardsProjectWeb.Content.StreamControllerTest do
  use ExUnit.Case, async: false
  use CardsProjectWeb.ConnCase

  @valid_params %{
    "title" => "test",
    "stream_url" => "https://example.com",
    "is_official" => true,
    "viewer_count_peak" => 0,
    "scheduled_start" => ~N[2024-01-01 00:00:00],
    "status" => "Scheduled",
    "platform" => "Twitch",
    "language" => "EN"
  }

  describe "GET /api/streams" do
    test "returns 200 with list", %{conn: conn} do
      conn = get(conn, "/api/streams")
      assert json_response(conn, 200) |> is_list()
    end
  end

  describe "GET /api/streams?q=test" do
    test "returns 200 with search results", %{conn: conn} do
      conn = get(conn, "/api/streams?q=test")
      assert json_response(conn, 200) |> is_list()
    end
  end

  describe "POST /api/streams" do
    test "creates record and returns 201", %{conn: conn} do
      conn = post(conn, "/api/streams", @valid_params)
      assert %{"id" => _id} = json_response(conn, 201)
    end
    test "fails when viewer_count_not_negative violated", %{conn: conn} do
      params = Map.put(@valid_params, "viewer_count_peak", -1)
      conn = post(conn, "/api/streams", params)
      assert conn.status in [400, 422]
    end
  end

  describe "GET /api/streams/:id" do
    test "returns 200 for existing record", %{conn: conn} do
      {:ok, record} = CardsProject.Content.create_stream(@valid_params)
      conn = get(conn, "/api/streams/#{record.id}")
      assert json_response(conn, 200)
    end
  end

  describe "PUT /api/streams/:id" do
    test "updates and returns 200", %{conn: conn} do
      {:ok, record} = CardsProject.Content.create_stream(@valid_params)
      conn = put(conn, "/api/streams/#{record.id}", @valid_params)
      assert json_response(conn, 200)
    end
  end


  describe "PATCH /api/streams/:id/transitions/scheduled-to-live" do
    test "transitions Scheduled -> Live", %{conn: conn} do
      {:ok, record} = CardsProject.Content.create_stream(@valid_params)
      conn = assign(conn, :current_user, %{role: "Streamer"})
      conn = patch(conn, "/api/streams/#{record.id}/transitions/scheduled-to-live")
      assert conn.status in [200, 409, 422, 404]
    end

    test "is forbidden with 403 for wrong role", %{conn: conn} do
      {:ok, record} = CardsProject.Content.create_stream(@valid_params)
      conn = assign(conn, :current_user, %{role: "admin"})
      conn = patch(conn, "/api/streams/#{record.id}/transitions/scheduled-to-live")
      assert conn.status == 403
    end
  end

  describe "PATCH /api/streams/:id/transitions/live-to-ended" do
    test "transitions Live -> Ended", %{conn: conn} do
      {:ok, record} = CardsProject.Content.create_stream(@valid_params)
      conn = assign(conn, :current_user, %{role: "Streamer"})
      conn = patch(conn, "/api/streams/#{record.id}/transitions/live-to-ended")
      assert conn.status in [200, 409, 422, 404]
    end

    test "is forbidden with 403 for wrong role", %{conn: conn} do
      {:ok, record} = CardsProject.Content.create_stream(@valid_params)
      conn = assign(conn, :current_user, %{role: "admin"})
      conn = patch(conn, "/api/streams/#{record.id}/transitions/live-to-ended")
      assert conn.status == 403
    end
  end

  describe "PATCH /api/streams/:id/transitions/ended-to-live" do
    test "is denied with 409", %{conn: conn} do
      {:ok, record} = CardsProject.Content.create_stream(@valid_params)
      conn = patch(conn, "/api/streams/#{record.id}/transitions/ended-to-live")
      assert conn.status in [409, 404]
    end
  end

end
